syntax = "proto3";

package TCPMessage;

message UserDescription {
    string UserName = 1;
}

message AuthMessage {
    enum Action {
        LOGIN = 0;
        LOGOUT = 1;
        ERR = 3;
    }
    Action              action = 1;
    UserDescription     user = 2;  //User nickname
    int32               code = 5;  //Response code
}

message PingMessage {
    enum Action {
        PING = 0;
        PONG = 1;
    }
    Action      action = 1; // Ping/Pong
    int32       pingId = 2; // Ping UNIQUE ID
}

message FileMessage {
    enum Action {
        DOWNLOAD = 0;
        UPLOAD = 1;
        RENAME = 2;
        DEL = 3;
        LIST = 4;
        ERR = 5;
    }
    Action          action = 1;
    UserDescription user = 2;
    string          fileName = 3;
    string          fileExtension = 4;
    uint64          fileSize = 5;
    int32           code = 6;
}

message TCPPacket {
    enum Type {
        AUTH = 0;
        PING = 1;
        FILE = 2;
        UNKNOWN = 3;
    }
    Type        type = 1;
    AuthMessage authMessage = 2;
    PingMessage pingMessage = 3;
    FileMessage fileMessage = 4;
}